File: 17-missing-rust/evidence/rust-hard-disk-parity-poc/hard_disk.go
package main

import (
    "fmt"
)

type Rat struct { n, d int64 }
func abs(x int64) int64 { if x < 0 { return -x }; return x }
func gcd(a,b int64) int64 { a=abs(a); b=abs(b); for b!=0 { a,b=b,a%b }; if a==0 { return 1 }; return a }
func R(n,d int64) Rat { if d==0 { panic("zero denominator") }; if d<0 { n=-n; d=-d }; g:=gcd(n,d); return Rat{n/g,d/g} }
func (a Rat) Add(b Rat) Rat { return R(a.n*b.d+b.n*a.d, a.d*b.d) }
func (a Rat) Sub(b Rat) Rat { return R(a.n*b.d-b.n*a.d, a.d*b.d) }
func (a Rat) Mul(b Rat) Rat { return R(a.n*b.n, a.d*b.d) }
func (a Rat) Div(b Rat) Rat { if b.n==0 { panic("division zero") }; return R(a.n*b.d,a.d*b.n) }
func (a Rat) Neg() Rat { return Rat{-a.n,a.d} }
func (a Rat) Cmp(b Rat) int { z:=a.n*b.d-b.n*a.d; if z<0{return -1}; if z>0{return 1};return 0 }
func (a Rat) String() string { if a.d==1{return fmt.Sprintf("%d",a.n)}; return fmt.Sprintf("%d/%d",a.n,a.d) }

type Disk struct { x,v Rat }
type State struct { a,b Disk; time Rat }
const ( 
    FNVOffset uint64 = 14695981039346656037
    FNVPrime uint64 = 1099511628211
)
func fnv(h uint64, s string) uint64 { for i:=0;i<len(s);i++ { h^=uint64(s[i]); h*=FNVPrime }; return h }
func encode(i int, kind string, dt Rat, s State) string { return fmt.Sprintf("%d|%s|%s|%s|%s|%s|%s\n",i,kind,dt.String(),s.time.String(),s.a.x.String(),s.a.v.String(),s.b.x.String()+","+s.b.v.String()) }
func wallTime(d Disk, lo, hi Rat) (Rat,bool) { if d.v.Cmp(R(0,1))>0{return hi.Sub(d.x).Div(d.v),true}; if d.v.Cmp(R(0,1))<0{return lo.Sub(d.x).Div(d.v),true}; return R(0,1),false }
func pairTime(s State, diameter Rat) (Rat,bool) { gap:=s.b.x.Sub(s.a.x).Sub(diameter); rel:=s.a.v.Sub(s.b.v); if rel.Cmp(R(0,1))<=0{return R(0,1),false}; t:=gap.Div(rel); return t,t.Cmp(R(0,1))>0 }
func advance(s *State, dt Rat) { s.a.x=s.a.x.Add(s.a.v.Mul(dt)); s.b.x=s.b.x.Add(s.b.v.Mul(dt)); s.time=s.time.Add(dt) }
func main() {
    // Equal hard disks of radius 1/2 in [0,8] x [0,3], constrained to y=3/2.
    // This is an exact 1D invariant subfamily of hard-disk billiards.
    lo,hi,diameter:=R(1,2),R(15,2),R(1,1)
    s:=State{Disk{R(1,1),R(1,1)},Disk{R(6,1),R(-2,1)},R(0,1)}
    h:=uint64(FNVOffset)
    const N=100000
    for i:=0;i<N;i++ {
        ta,oka:=wallTime(s.a,lo,hi); tb,okb:=wallTime(s.b,lo,hi); tp,okp:=pairTime(s,diameter)
        kind:=""; dt:=R(0,1)
        if okp && (!oka || tp.Cmp(ta)<0) && (!okb || tp.Cmp(tb)<0) { kind="PAIR"; dt=tp
        } else if oka && (!okb || ta.Cmp(tb)<0) { kind="A_WALL"; dt=ta
        } else if okb && (!oka || tb.Cmp(ta)<0) { kind="B_WALL"; dt=tb
        } else { panic("tie or no event") }
        advance(&s,dt)
        switch kind { case "PAIR": s.a.v,s.b.v=s.b.v,s.a.v; case "A_WALL": s.a.v=s.a.v.Neg(); case "B_WALL": s.b.v=s.b.v.Neg() }
        row:=encode(i+1,kind,dt,s); h=fnv(h,row)
        if i<12 { fmt.Print(row) }
    }
    fmt.Printf("SUMMARY events=%d time=%s ax=%s av=%s bx=%s bv=%s fnv64=%016x\n",N,s.time.String(),s.a.x.String(),s.a.v.String(),s.b.x.String(),s.b.v.String(),h)
}

File: 17-missing-rust/evidence/rust-hard-disk-parity-poc/hard_disk.rs
use std::fmt;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
struct Rat { n: i128, d: i128 }
impl Rat {
    fn new(mut n:i128, mut d:i128)->Self { assert!(d!=0); if d<0 {n=-n;d=-d}; let g=gcd(n,d); Self{n:n/g,d:d/g} }
    fn add(self,b:Self)->Self { Self::new(self.n*b.d+b.n*self.d,self.d*b.d) }
    fn sub(self,b:Self)->Self { Self::new(self.n*b.d-b.n*self.d,self.d*b.d) }
    fn mul(self,b:Self)->Self { Self::new(self.n*b.n,self.d*b.d) }
    fn div(self,b:Self)->Self { assert!(b.n!=0); Self::new(self.n*b.d,self.d*b.n) }
    fn neg(self)->Self { Self{n:-self.n,d:self.d} }
    fn cmp(self,b:Self)->i8 { let z=self.n*b.d-b.n*self.d; if z<0{-1}else if z>0{1}else{0} }
}
fn gcd(mut a:i128,mut b:i128)->i128 { a=a.abs(); b=b.abs(); while b!=0 { let t=a%b;a=b;b=t }; if a==0{1}else{a} }
impl fmt::Display for Rat { fn fmt(&self,f:&mut fmt::Formatter<'_>)->fmt::Result { if self.d==1 {write!(f,"{}",self.n)} else {write!(f,"{}/{}",self.n,self.d)} } }
#[derive(Clone,Copy)] struct Disk { x:Rat, v:Rat }
#[derive(Clone,Copy)] struct State { a:Disk, b:Disk, time:Rat }
const FNV_OFFSET:u64=14695981039346656037; const FNV_PRIME:u64=1099511628211;
fn fnv(mut h:u64,s:&str)->u64 { for b in s.bytes(){h^=b as u64;h=h.wrapping_mul(FNV_PRIME);}h }
fn wall_time(d:Disk,lo:Rat,hi:Rat)->Option<Rat>{ if d.v.cmp(Rat::new(0,1))>0 {Some(hi.sub(d.x).div(d.v))} else if d.v.cmp(Rat::new(0,1))<0 {Some(lo.sub(d.x).div(d.v))} else {None} }
fn pair_time(s:State,diameter:Rat)->Option<Rat>{ let gap=s.b.x.sub(s.a.x).sub(diameter); let rel=s.a.v.sub(s.b.v); if rel.cmp(Rat::new(0,1))<=0{return None}; let t=gap.div(rel); if t.cmp(Rat::new(0,1))>0{Some(t)}else{None} }
fn advance(s:&mut State,dt:Rat){s.a.x=s.a.x.add(s.a.v.mul(dt));s.b.x=s.b.x.add(s.b.v.mul(dt));s.time=s.time.add(dt);}
fn main(){
    // Equal hard disks of radius 1/2 in [0,8] x [0,3], constrained to y=3/2.
    // Exact 1D invariant subfamily of hard-disk billiards.
    let (lo,hi,diameter)=(Rat::new(1,2),Rat::new(15,2),Rat::new(1,1));
    let mut s=State{a:Disk{x:Rat::new(1,1),v:Rat::new(1,1)},b:Disk{x:Rat::new(6,1),v:Rat::new(-2,1)},time:Rat::new(0,1)};
    let mut h=FNV_OFFSET;
    const N:usize=100000;
    for i in 0..N {
        let ta=wall_time(s.a,lo,hi); let tb=wall_time(s.b,lo,hi); let tp=pair_time(s,diameter);
        let (kind,dt)=match (tp,ta,tb) {
            (Some(p),a,b) if a.map_or(true,|x|p.cmp(x)<0) && b.map_or(true,|x|p.cmp(x)<0) => ("PAIR",p),
            (_,Some(a),b) if b.map_or(true,|x|a.cmp(x)<0) => ("A_WALL",a),
            (_,_,Some(b)) => ("B_WALL",b),
            _ => panic!("tie or no event"),
        };
        advance(&mut s,dt);
        match kind { "PAIR"=>{let v=s.a.v;s.a.v=s.b.v;s.b.v=v;}, "A_WALL"=>s.a.v=s.a.v.neg(), "B_WALL"=>s.b.v=s.b.v.neg(), _=>unreachable!() }
        let row=format!("{}|{}|{}|{}|{}|{}|{},{}\n",i+1,kind,dt,s.time,s.a.x,s.a.v,s.b.x,s.b.v);
        h=fnv(h,&row);
        if i<12 { print!("{}",row); }
    }
    println!("SUMMARY events={} time={} ax={} av={} bx={} bv={} fnv64={:016x}",N,s.time,s.a.x,s.a.v,s.b.x,s.b.v,h);
}
